In [5]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [6]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [7]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [8]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [9]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
columns=df_market_data.columns
df_market_data_scaled = StandardScaler().fit_transform(df_market_data)

In [10]:
# Create a DataFrame with the scaled data
df_market_data_scale2 = pd.DataFrame(df_market_data_scaled, columns= df_market_data.columns)

# Copy the crypto names from the original DataFrame
df_market_data_scale2["coin_id"]= df_market_data.index

# Set the coin_id column as index
df_market_data_scale2 = df_market_data_scale2.set_index("coin_id")


# Display the scaled DataFrame
df_market_data_scale2


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [11]:
# Create a list with the number of k-values from 1 to 11
inertia = []


In [12]:
# Create an empty list to store the inertia values
k= list(range(1,11))

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model= KMeans(n_clusters=i, random_state= 1)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)


C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Window

In [13]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)


In [14]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k",
    y="inertia",
    title= "Elbow Curve",
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  The best value for 'k' is 4.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [15]:
# Initialize the K-Means model using the best value for k
model= KMeans(n_clusters=4)

# Fit the K-Means model using the scaled DataFrame
model.fit(df_market_data_scale2)

In [16]:
model.fit(df_market_data_scale2)

C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4)

In [17]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
c_cluster=model.predict(df_market_data_scale2)

# Print the resulting array of cluster values.
c_cluster

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 2, 0, 1, 1, 3, 1, 1, 1, 1])

In [18]:
# Create a copy of the scaled DataFrame
df_copy=df_market_data_scale2.copy()
df_copy

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [19]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_copy["Predicted"]=c_cluster

# Display the copy of the scaled DataFrame
df_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Predicted  
coin_id                                              
bitcoin                        -0.251637          0  
ethereum                       -0.199352          0  
tether                         -0.282061          1  
ripple                         -0.295546          1  
bitcoin-cash                   -0.270317          0

In [20]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
df_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Predicted",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by CryptoCurrency - k=4"
)

:NdOverlay   [Predicted]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [21]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [22]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
df_copy_pca_data = pca.fit_transform(df_copy)

# View the scaled PCA data
df_copy_pca_data[:5]

array([[-0.82258702,  0.83270555,  0.56655379],
       [-0.70990347,  0.45421229,  1.05640205],
       [-0.30789987, -0.18460919, -0.74081973],
       [-0.353704  , -0.23981488, -0.59614094],
       [-1.48496628,  2.01848451,  1.78701064]])

In [23]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3738947 , 0.32920008, 0.1850154 ])

In [24]:
sum(pca.explained_variance_ratio_)

0.8881101751527735

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is 0.8881101751527735.

In [25]:
# Create a new DataFrame with the PCA data.
df_new_pca_data = pd.DataFrame(df_copy_pca_data, columns=["PC1", "PC2", "PC3"])

# Copy the crypto names from the original scaled DataFrame
df_new_pca_data["Crypto Name"]= df_market_data.index

# Set the coin_id column as index
df_new_pca_data= df_new_pca_data.set_index("Crypto Name")

# Display the scaled PCA DataFrame
df_new_pca_data

PC1       PC2       PC3
Crypto Name                                       
bitcoin              -0.822587  0.832706  0.566554
ethereum             -0.709903  0.454212  1.056402
tether               -0.307900 -0.184609 -0.740820
ripple               -0.353704 -0.239815 -0.596141
bitcoin-cash         -1.484966  2.018485  1.787011
binancecoin          -0.781024  1.381609  0.902062
chainlink            -0.833238  0.519208  2.839653
cardano              -0.644167  0.729885  1.565143
litecoin             -0.868642  0.421293  0.694079
bitcoin-cash-sv      -0.649172 -0.225935 -0.392029
crypto-com-chain     -0.040712 -1.389292 -1.479353
usd-coin             -0.311314 -0.192038 -0.762095
eos                  -0.556771 -0.497469 -0.674255
monero               -0.310518  2.920727  1.624866
tron                 -0.314660 -0.121673 -0.141549
tezos                -1.019695 -0.507678  1.136937
okb                   0.229683 -1.272646 -1.084205
stellar              -0.386659 -0.748870 -0.198786
cosmos               -0.623928  0.705798  1.799430
cdai                 -0.384825 -0.161690 -0.767607
neo                  -0.214733 -1.001127 -0.798546
wrapped-bitcoin      -0.824071  0.817177  0.545012
leo-token            -0.246613 -0.691052 -1.141461
huobi-token          -0.302089 -0.227363 -0.462716
nem                   0.709842  0.575616 -1.064154
binance-usd          -0.324386 -0.168029 -0.749075
iota                 -0.676445 -0.541351  0.011075
vechain              -0.338151 -1.959313 -1.327981
zcash                -0.686153  0.489983  1.160207
theta-token           2.773629  0.056124 -1.521206
dash                 -0.541847 -0.498020  0.161132
ethereum-classic     -0.480483 -0.375452 -0.264970
ethlend               8.064371 -3.664118  2.999038
maker                -0.597629  0.161290  0.532155
havven                0.933272 -2.236852  0.340675
omisego               0.237903  0.425833 -1.189850
celsius-degree-token  5.153744  6.878652 -1.924068
ontology             -0.450823 -2.129476 -0.758038
ftx-token            -0.526103  0.002803  0.038193
true-usd             -0.333235 -0.152945 -0.738993
digibyte             -0.155298 -0.204590 -0.981723

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [26]:
# Create a list with the number of k-values from 1 to 11
k2= list(range(1,11))

In [27]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k2:
    k_model= KMeans(n_clusters=i, random_state= 1)
    k_model.fit(df_new_pca_data)
    inertia_pca.append(k_model.inertia_)
    
inertia_pca

C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Window

[268.70748104317477,
 189.18904573376187,
 133.41998713480584,
 46.685148434927974,
 36.86413603457518,
 26.53590730454502,
 22.448626220280705,
 16.758927704069905,
 13.729223713376292,
 11.450690204180452]

In [28]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow={
    "k2":k2,
    "inertia2": inertia_pca
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow2 = pd.DataFrame(pca_elbow)

df_elbow2

k2    inertia2
0   1  268.707481
1   2  189.189046
2   3  133.419987
3   4   46.685148
4   5   36.864136
5   6   26.535907
6   7   22.448626
7   8   16.758928
8   9   13.729224
9  10   11.450690

In [29]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow2.hvplot.line(
    x="k2",
    y="inertia2",
    title= "Elbow Curve",
    xticks=k2
)

:Curve   [k2]   (inertia2)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for "k" using PCA data is 3.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, it is a difference from the k value in the original data.
  * 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [30]:
# Initialize the K-Means model using the best value for k
model2= KMeans(n_clusters=3)

In [31]:
# Fit the K-Means model using the PCA data
model2.fit(df_new_pca_data)

C:\Users\letni\Anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=3)

In [32]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_clusters= model2.predict(df_new_pca_data)

# Print the resulting array of cluster values.
print(pca_clusters)

[1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 0 1 0 0 2
 0 0 0 0]


In [33]:
# Create a copy of the scaled PCA DataFrame
df_new_pca_data_predict2= df_new_pca_data.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_new_pca_data_predict2["PCAClusters"]= pca_clusters

# Display the copy of the scaled PCA DataFrame
df_new_pca_data_predict2.head()

PC1       PC2       PC3  PCAClusters
Crypto Name                                            
bitcoin      -0.822587  0.832706  0.566554            1
ethereum     -0.709903  0.454212  1.056402            1
tether       -0.307900 -0.184609 -0.740820            0
ripple       -0.353704 -0.239815 -0.596141            0
bitcoin-cash -1.484966  2.018485  1.787011            1

In [34]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
df_new_pca_data_predict2.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="PCAClusters",
    title = "Scatter Plot by PCAClusters"
)

:NdOverlay   [PCAClusters]
   :Scatter   [PC1]   (PC2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [35]:
# Composite plot to contrast the Elbow curves
elbow_plot_1 = df_elbow.hvplot.line(
    x="k",
    y="inertia",
    title= "Elbow Curve",
    xticks=k
)
elbow_plot_2=df_elbow2.hvplot.line(
    x="k2",
    y="inertia2",
    title= "Elbow Curve",
    xticks=k2
)
combined_elbow_plots= elbow_plot_1 + elbow_plot_2

combined_elbow_plots


:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k2]   (inertia2)

In [36]:
# Composite plot to contrast the clusters
# Code is below

In [37]:
scatter_plot_1 = df_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Predicted",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by CryptoCurrency - k=4")
scatter_plot_2 = df_new_pca_data_predict2.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="PCAClusters",
    title = "Scatter Plot by PCAClusters"
)
combined_scatter_plots= scatter_plot_1 + scatter_plot_2

combined_scatter_plots


:Layout
   .NdOverlay.I  :NdOverlay   [Predicted]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [PCAClusters]
      :Scatter   [PC1]   (PC2)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The impact of using fewer features to cluster the data with KMeans is that the cluster is bunched together and not really displaying a clear picture. 